File: LeetCode/216组合总和Ⅲ.java
class Solution {
    List<List<Integer>> ans;
    LinkedList<Integer> path;
    public List<List<Integer>> combinationSum3(int k, int n) {
        ans = new ArrayList<>();
        path = new LinkedList<>();
        backtracking(1, 0, k, n);
        return ans;
    }

    public void backtracking(int left, int sum, int k, int n) {
        if (sum > n) return ;
        if (path.size() == k && sum == n) {
            ans.add(new ArrayList<>(path));
            return ;
        }
        for (int i = left; i <= 9 - (k - path.size() - 1); i++) {
            sum += i;
            path.add(i);
            backtracking(i + 1, sum, k, n);
            path.removeLast();
            sum -= i;
        }
        return ;
    }
}

File: LeetCode/216组合总和Ⅲ.go
var ans [][]int

func combinationSum3(k int, n int) [][]int {
	ans = [][]int{}
	backtracking(1, 0, k, n, []int{})
	return ans
}

func backtracking(left int, sum int, k int, n int, path []int) {
	if len(path) == k && sum == n {
		temp := make([]int, k)
		copy(temp, path)
		ans = append(ans, temp)
		return
	}
	for i := left; i <= 9; i++ {
		sum += i
		path = append(path, i)
		backtracking(i+1, sum, k, n, path)
		path = path[:len(path)-1]
		sum -= i
	}
	return
}	